File: frontend/src/services/salesService.js
import api from './api';

const salesService = {
  // Existing methods...
  getSalesHistory: async (params) => {
    const response = await api.get('/sales/history', { params });
    return response.data;
  },

  downloadInvoice: async (saleId) => {
    const response = await api.get(`/sales/${saleId}/invoice`, {
      responseType: 'blob',
    });
    
    // Create download link
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement('a');
    link.href = url;
    link.setAttribute('download', `invoice-${saleId}.pdf`);
    document.body.appendChild(link);
    link.click();
    link.remove();
  },

  createSale: async (saleData) => {
    const response = await api.post('/sales/create', saleData);
    return response.data;
  },

  getMedicineSalesStats: async (params) => {
    // Filter out empty values
    const cleanParams = {};
    if (params.startDate) cleanParams.startDate = params.startDate;
    if (params.endDate) cleanParams.endDate = params.endDate;
    
    const response = await api.get('/sales/medicine-stats', { params: cleanParams });
    return response.data;
  },

  cancelSale: async (saleId) => {
    const response = await api.post(`/sales/${saleId}/cancel`);
    return response.data;
  },
};

export default salesService;

File: frontend1/lib/services/sales_service.dart
import 'package:dio/dio.dart';
import 'package:frontend1/models/sale.dart';
import 'package:frontend1/services/api_service.dart';

class SalesService {
  final ApiService _apiService = ApiService();

  /// Créer une nouvelle vente
  Future<Map<String, dynamic>> createSale(Map<String, dynamic> saleData) async {
    try {
      final response = await _apiService.post(
        '/sales/create',
        data: saleData,
      );
      return response.data as Map<String, dynamic>;
    } catch (e) {
      print('[SalesService] Erreur createSale: $e');
      rethrow;
    }
  }

  /// Annuler une vente
  Future<void> cancelSale(int saleId) async {
    try {
      await _apiService.post('/sales/$saleId/cancel');
    } catch (e) {
      print('[SalesService] Erreur cancelSale: $e');
      rethrow;
    }
  }

  /// Télécharger une facture (PDF)
  /// Retourne les bytes du PDF
  Future<List<int>> downloadInvoice(int saleId) async {
    try {
      final response = await _apiService.get(
        '/sales/$saleId/invoice',
        options: Options(
          responseType: ResponseType.bytes,
        ),
      );
      return response.data as List<int>;
    } catch (e) {
      print('[SalesService] Erreur downloadInvoice: $e');
      rethrow;
    }
  }

  /// Récupérer l'historique des ventes
  Future<SalesHistoryResponse> getSalesHistory({
    int page = 1,
    String? search,
    String? startDate,
    String? endDate,
    double? minAmount,
    double? maxAmount,
  }) async {
    try {
      final queryParams = <String, dynamic>{'page': page, 'page_size': 20};
      if (search != null) queryParams['search'] = search;
      if (startDate != null && startDate.isNotEmpty) queryParams['start_date'] = startDate;
      if (endDate != null && endDate.isNotEmpty) queryParams['end_date'] = endDate;
      if (minAmount != null) queryParams['min_amount'] = minAmount;
      if (maxAmount != null) queryParams['max_amount'] = maxAmount;

      final response = await _apiService.get(
        '/sales/history',
        queryParameters: queryParams,
      );
      return SalesHistoryResponse.fromJson(response.data);
    } catch (e) {
      print('[SalesService] Erreur getSalesHistory: $e');
      rethrow;
    }
  }

  /// Récupérer les statistiques de ventes par médicament
  Future<List<MedicineSaleStats>> getMedicineSalesStats({
    String? startDate,
    String? endDate,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (startDate != null && startDate.isNotEmpty) queryParams['start_date'] = startDate;
      if (endDate != null && endDate.isNotEmpty) queryParams['end_date'] = endDate;

      final response = await _apiService.get(
        '/sales/medicine-stats',
        queryParameters: queryParams,
      );
      
      return (response.data as List)
          .map((item) => MedicineSaleStats.fromJson(item))
          .toList();
    } catch (e) {
      print('[SalesService] Erreur getMedicineSalesStats: $e');
      rethrow;
    }
  }
}
